src/nfa/operations.cc

```diff
@@ -56,6 +56,126 @@ namespace {
         return lts_for_simulation.compute_simulation();
     }
 
+    bool vec_contains(std::vector<std::pair<State, State>> vec, std::pair<State, State> par){
+        for (size_t i = 0; i < vec.size(); i++){
+            if (vec[i] == par){
+                return true;
+            }
+        }
+        return false;
+    }
+
+    Nfa reduce_testing(const Nfa& aut){
+        // ! Preprocessing
+        Nfa result_nfa;
+        Nfa reverted_nfa;
+
+        std::vector<std::pair<State, State>> result_sim {}; // R
+        std::vector<std::pair<State, State>> worklist {}; // Worklist
+
+        // Alphabet extraction
+        mata::OnTheFlyAlphabet alph;
+        aut.fill_alphabet(alph);
+        std::vector<Symbol> alph_syms = alph.get_alphabet_symbols().to_vector();
+
+        size_t no_states = aut.num_of_states();
+        size_t ***matrix;
+
+        // Matrix allocation
+        matrix = static_cast<size_t ***>(malloc(sizeof(size_t **) * alph_syms.size()));
```

Comment:
1) no `malloc/free` in C++, better use `new/delete`
2) anyway, the best is to completely avoid dynamic allocation and just use a `std::vector`
3) also, no need to do nested arrays, just use one-dimensional and write index functions to access the right cell
4) `size_t` probably an overkill, a 32b counter should be enough

Reply:
Using one dimensional array now.